### ISSUE TITLE
loop optimizations

### ISSUE THREAD

**Issue by David10238** [FIRST-Tech-Challenge/SkyStone#3]
I have a couple of questions and requests.
Doing some benchmarks on the control hub the community suspected caching writes was added at some point of time. I looked at LynxDcMotorController and confirmed this. Checking the caching algorithm in LastKnown, I noticed msFreshness, which forces a write after half a second. Since this only really matters after brownouts and DCs, which are incredibly rare for a top-level team, I would like to have the ability to easily either disable msFreshness or set it to a higher value. You can actually notice the effect of the writes every half second in a benchmark.
Another thing with the control hub they noticed was [issue 542](uhttps://github.com/ftctechnh/ftc_app/issues/542rl) appeared to be fixed. Was anything done that should cause this?
Bulk took < 2/3s of the time and non-bulk interactions like setting motor powers or reading encoders took < 1/2 the time. Since it was announced the control hub has no competitive advantage I am wondering if there was an sdk change that optimized the refresh rate.
I plan on rebenchmarking the ExH with the same code this afternoon to give a fair comparison. I'm not in a pilot region, so I can only do limited testing by asking people I know to do testing on control hub.

The main request the community has is bulk writes. This would take the already insane 227Hz refresh rate of one bulk read and 4 motor writes on control hub to the next level.

**Comment by cmacfarl** [comment]
> You can actually notice the effect of the writes every half second in a benchmark.

What does notice mean?  Can you actually quantify this with data?

**Comment by cmacfarl** [comment]
> Another thing with the control hub they noticed was issue 542 appeared to be fixed. Was anything done that should cause this?
Bulk took 2/3s of the time and non-bulk interactions like setting motor powers or reading encoders took 1/2 the time. 

Regarding the i2c transaction times can you post here the firmware version of both the expansion hub and control hub used to do the comparison?

**Comment by gearsincorg** [comment]
@David10238  I'll be doing some of my own benchmarking on the new Control Hub shortly.  If you are interested in sharing any of your procedures/results' I would be interested in comparing/combining them with my own results.

Note that all repeated device (servo/motor) writes of the same value are essentially ignored, so for meaningful results I always ensure that the written value is changing significantly to force a fresh write.  I figure that motors being controlled by PID loops invariably never get sent the same value twice in a row, unless they are stopped,

**Comment by David10238** [comment]
Initial testing of the control hub showed ~1.2 ms hardware call, ~2 ms bulk read, ~7-8 ms. When we changed the software to average over a longer time, the times went down very slightly. With ExH it would take ~3ms for any non-2C hardware call. https://pastebin.com/LLXSexMG. The AveragingCircleBuffer only averages the last 5000 inputs, which would take about 20 seconds to cycle through with just over 200Hz. 
![image](https://user-images.githubusercontent.com/28957979/61470498-d4880680-a946-11e9-9cef-5b94c683c06d.png)
This is the results ran on a control hub. The code was the same as issue 542's benchmark with minor edits to work with Java 7 (OnBot). I don't have the edited code that was run though.

**Comment by rbrott** [comment]
> Another thing with the control hub they noticed was issue 542 appeared to be fixed. Was anything done that should cause this?

The data you posted doesn't seem to support this conclusion. 542 is unlikely to be fixed until either a firmware update or modification of the read window mechanism.

**Comment by cmacfarl** [comment]
This shows that the control hub's firmware is 1.8.2, which is what I'd
expect, but the other question I had was what was the firmware of the
expansion hub you were comparing against?

**Comment by FROGbots-4634** [comment]
> The main request the community has is bulk writes. This would take the already insane 227Hz refresh rate of one bulk read and 4 motor writes on control hub to the next level.

Actually a "bulk write" isn't what you want. What you really want is a "Bulk Data Exchange" command, where the initial command would essentially be a "bulk write", but then the ACK that is returned for it also contains all of the data that a "bulk read" returns. This would allow a "bulk read" and a "bulk write" in ONE command. Note that the SDK and firmware already support this type of command architecture - implementing this would be very easy.

**Comment by FROGbots-4634** [comment]
> Since it was announced the control hub has no competitive advantage I am wondering if there was an sdk change that optimized the refresh rate.

The SBC in the Control Hub does not communicate over USB+FTDI to the Lynx PCB. Rather, it is directly attached through UART. This is likely the cause of the faster hardware calls on the Control Hub.

**Comment by AlecHub** [comment]
>  Since it was announced the control hub has no competitive advantage ...

If you believe the control hub has no competitive advantage, I have some perfectly legal NXTs for sale.

**Comment by FROGbots-4634** [comment]
> If you believe the control hub has no competitive advantage, I have some perfectly legal NXTs for sale.

:rofl:

However, the Control Hub is also a bit of a disadvantage in that it has no screen, which can make understanding what's going on very difficult. Looking at camera feed for instance would require hooking up an HDMI monitor. Also, very occasionally I've had the SDK get into what I assume must be deadlock, which could only be solved by closing and reopening the app on the RC.

**Comment by AlecHub** [comment]
> However, the Control Hub is also a bit of a disadvantage in that it has no screen, …

The NXT has a stunning 6400 pixel screen, plus integrated D/A converter, high def speaker, and user friendly keyboard, all in an attractive bullet-proof case.  The NXT is a much better value than the control hub.

**Comment by FROGbots-4634** [comment]
> > However, the Control Hub is also a bit of a disadvantage in that it has no screen, …
> 
> The NXT has a stunning 6K pixel screen, plus integrated D/A converter, high def speaker, and user friendly keyboard, all in an attractive bullet-proof case. The NXT is a much better value than the control hub. I have a limited supply of these bad boys so act fast.

The Control Hub is a tradeoff - no screen, only 1GB RAM and a slower CPU than in the phones (not necessarily a problem for robot control tasks although computer vision performance will likely suffer a bit) and having to wait for it to boot each time power is lost - but gaining a much more compact setup with less wires and likely a good bit more ESD resiliency due to the internal UART link to the Lynx, and since the ROM is developed in-house you're unlikely to run into weird bugs like the Motorola WiFi scanning bug that hit this year.

**Comment by AlecHub** [comment]
> The Control Hub is a tradeoff - no screen, only 1GB RAM and a slower CPU than in the phones (not necessarily a problem for robot control tasks although computer vision performance will likely suffer a bit) …

The [Control Hub wiki](../wiki/Managing-a-Control-Hub) states **"the Control Hub supports both the 2.4 GHz and 5 GHz"** WiFi bands.  However the [DragonBoard 410c](https://developer.qualcomm.com/hardware/dragonboard-410c) supports only the 2.4 GHz band.  This suggests that the SBC on the control hub has been upgraded to a [DragonBoard 820c](https://developer.qualcomm.com/hardware/dragonboard-820c) or better. The 820c has 3GB RAM (1866 MHz) and a quad core processor running at up to 2.35GHz per core.  Plus the 820c has a "Qualcomm® Hexagon™ 680 DSP" for vision acceleration.

**Comment by FROGbots-4634** [comment]
> This suggests that the SBC on the control hub has been upgraded to a [DragonBoard 820c](https://developer.qualcomm.com/hardware/dragonboard-820c) or better.

Nope. It has been switched to REV's fully custom SBC which is based off a Rockchip SOC that is about equivalent to the SD410 and still has 1GB of RAM.

**Comment by AlecHub** [comment]
> Nope. It has been switched to REV's fully custom SBC which is based off a Rockchip SOC that is about equivalent to the SD410 and still has 1GB of RAM.

Good to know, thanks!
